Governor Documentation and Model

lukelowry opened the pull request:
## Description
 
Documentation for the phasor dynamics governor (TGOV1) that will be implemented following this PR.
 
 
 ## Proposed changes
 
Restructured folder to allow for many gov. types and added documentation for TGOV1
 
 
 ## Further comments
 
There are piece-wise differential equations in this system. I wrote a short proposal on how to implement this to cooperate with the general solver. Please let me know if a bump function is sufficient or if the general solver can handle piece-wise definition (not to my knowledge).

Also there was a small latex issue in GENROU that I included in this PR.

 ## Checklist
 
- [x] Governor Implementation
- [ ] Proper CMake build configuration.
- [x] Documentation in a form of README file with model equations.
- [ ] Example of usage.
- [ ] All tests pass.
- [x] The new code follows GridKit™ style guidelines.
- [ ] There are unit tests for the new code.
- [ ] The feature branch is rebased with respect to the target branch.

lukelowry committed:
init: governor README folder structure

lukelowry committed:
init: governor README folder structure

lukelowry committed:
edit: equations but misplaced

lukelowry committed:
edit: equations but misplaced

lukelowry committed:
edit: variables and parameters

lukelowry committed:
edit: variables and parameters

lukelowry committed:
edit: alg and diff equations

lukelowry committed:
edit: alg and diff equations

lukelowry committed:
edit: differential system

lukelowry committed:
edit: differential system

lukelowry committed:
edit: domain of piecewise

lukelowry committed:
edit: domain of piecewise

lukelowry committed:
edit: notation cleanup

lukelowry committed:
edit: notation cleanup

lukelowry committed:
propose: continuous domain constraint

lukelowry committed:
propose: continuous domain constraint

lukelowry committed:
edit: bump function

lukelowry committed:
edit: bump function

lukelowry committed:
edit: syntax error

lukelowry committed:
edit: syntax error

lukelowry committed:
edit: completed TGOV1 draft

lukelowry committed:
edit: completed TGOV1 draft

lukelowry committed:
Apply pre-commmit fixes

lukelowry committed:
Apply pre-commmit fixes

pelesh committed:
Update developer guidelines with Doxygen instructions (#99)

pelesh committed:
Reorganize examples by model family (#104)

pelesh commented on src/Model/PhasorDynamics/Governor/TGOV1/README.md:
```diff
@@ -0,0 +1,80 @@
+# **TGOV1**
+
```
Probably a good idea to spell out this acronym.

lukelowry replied:
It is now **_Steam Turbine-Governor Model (TGOV1)_**

pelesh commented on src/Model/PhasorDynamics/Governor/TGOV1/README.md:
```diff
@@ -0,0 +1,80 @@
+Standard model of the stream turbine
+
+<div align="center">
+   <img align="center" src="../../../../../docs/Figures/TGOV1.JPG">
+   
```
Nor all variables from the equations show up in this graphics. For example, where is $P_{tx}$ input in this schematics?

lukelowry replied:
The differential variables $P_{tx}$ and $P_{v}$ are referenced to (1) and (2), respectively in the flow diagram.

pelesh commented on src/Model/PhasorDynamics/Governor/TGOV1/README.md:
```diff
@@ -0,0 +1,80 @@
+## Nomenclature
+
+### Algebraic Variables
+- $\omega$ - Per-unit Speed Deviation
+- $P_{ref}$ - Reference Power
+- $P_{mech}$ - Mechnical Power to Generator
```
Please specify units explicitly. Isn't $\omega$ expressed in [1/s] even in power units?

pelesh replied:
Also in the figure $\Delta \omega$ is used instead of $\omega$.

lukelowry replied:
No, p.u. frequency/speed is unitless, such that $\omega_{p.u.}=1$ represents nominal frequency. 

I will change $\omega \mapsto \Delta \omega$ since other models in the PhasorDynamics module use $\omega$ for speed, rather than speed deviation.

pelesh commented on src/Model/PhasorDynamics/Governor/TGOV1/README.md:
```diff
@@ -0,0 +1,80 @@
+- $P_{mech}$ - Mechnical Power to Generator
+
+### Differential Variables
+- $P_{tx}$ - Turbine Power 
+- $P_{v}$ - Valve Position
+
+### Parameters
+- $R$ - Droop Constant
+- $T_1$ - Valve Time Delay
+- $T_2$ - Turbine Numerator Time Constat
+- $T_3$ - Turbine Delay
+- $V_{max}$ - Max Valve Position
+- $V_{min}$ -  Min Valve Position
+- $D_t$ - Turbine Damping Coefficient
```
Units need to be specified. If units are dimensionless, that should be stated explicitly or expressed as [1].

pelesh replied:
Also $V_{max}$ and $V_{min}$ could be confused with voltages. Is this common notation? It would make more sense to use $P_{vmax}$ and $P_{vmin}$, respectively or even $\overline P_{v}$ and $\underline{P_{v}}$.

pelesh replied:
If I am not mistaken, label (2) in the figure denotes $P_v$ and label (1) denotes $P_{tx}$. If so, this has to be specified in the text.

Another question -- what is the input to the negative signal input to the summation on the left-hand-side of the diagram? It looks as if it is $\Delta \omega$, but if this is the case, units need to be explained.

lukelowry replied:
1.  Complete

2.   I agree here, I propose $P_{vmin}$ and $P_{vmax}$ as they are the clearest in my estimation.

3. Correct, this has now been specified under Nomenclature.  Also, the input you referenced is indeed $\Delta \omega$. Both  $\Delta \omega$ and $P_{ref}$ are in per-unit. The unit specification should clarify this now.

pelesh commented on src/Model/PhasorDynamics/Governor/TGOV1/README.md:
```diff
@@ -0,0 +1,80 @@
+   \dot{P}_{v}    &= 
+      \dfrac{1}{T_1}
+   \begin{cases}
+      -P_{v} + \frac{1}{R}(P_{ref}-\omega)
+         &  \text{if } P_v \in (V_{min}, V_{max})\\
```
Use `\dfrac` so that 1/R is more prominently displayed:
```suggestion
      -P_{v} + \dfrac{1}{R}(P_{ref}-\omega)
         &  \text{if } P_v \in (V_{min}, V_{max})\\
```

lukelowry replied:
Corrected

pelesh commented on src/Model/PhasorDynamics/Governor/TGOV1/README.md:
````diff
@@ -0,0 +1,80 @@
+following.
+```math
+\begin{aligned}
+   I(x,x_0,\epsilon) 
+            &= \dfrac{1}{2} + \dfrac{1}{2}\tanh 
+            \left(\dfrac{x-x_0}{\epsilon}\right) \\
+   \phi(x)  &= I \left(x;V_{min}, \epsilon \right) -
+            I \left(x;V_{max}, \epsilon \right)
+            \qquad \epsilon <<1 \\
+   \dot{P}_{v}   
+            &= 
+            \dfrac{1}{T_1}\cdot \phi(P_v)
+            \left[
+               -P_{v} + \frac{1}{R}(P_{ref}-\omega)
+            \right] \\
````
I am not sure this is correct approximation.

lukelowry replied:
I will generalize this for now until we converge on a solution for smooth indicator.

pelesh reviewed (changes requested):
Great work! It would be good to make some additional clarification and make sure equations match the signal flow diagram.

lukelowry committed:
edit: parameter names etc [skip ci]

lukelowry committed:
edit: parameter names etc [skip ci]

lukelowry committed:
Merge branch 'lukel/phasor_tgov1_dev' of https://github.com/ORNL/GridKit into lukel/phasor_tgov1_dev

lukelowry committed:
edit: delta omega syntax [skip ci]

lukelowry committed:
edit: delta omega syntax [skip ci]

lukelowry committed:
edit: power in p.u. [skip ci]

lukelowry committed:
edit: power in p.u. [skip ci]

lukelowry committed:
edit: indicator generalization [skip ci]

lukelowry committed:
edit: indicator generalization [skip ci]

lukelowry commented:
Comments addressed. Awaiting @abirchfield comments, especially in regards to clarifying units.

lukelowry committed:
add: untested model draft cpp [skip ci]

lukelowry committed:
add: untested model draft cpp [skip ci]

lukelowry commented:
Untested implementation in recent commit, aligned with documentation.

lukelowry committed:
todo: added dev comments [skip ci]

lukelowry committed:
todo: added dev comments [skip ci]

nkoukpaizan committed:
Fixes to assertions and overrides (#106)

* size_quad() --> sizeQuadrature().

* Add missing overrides in IdaTests.

* Fix time assert with tolerance within a namespace.

* Use GridKit's isEqual in example2 time stamp assertion.

---------

Co-authored-by: nkoukpaizan <nkoukpaizan@users.noreply.github.com>

nkoukpaizan committed:
Verify Branch and Load Jacobians computed with Sparse::Variable (#89)

---------

Co-authored-by: Alexander Novotny <alex_novo@vt.edu>

nkoukpaizan committed:
Fix load Jacobian (#108)

* Fix load Jacobian.

* Upgrade base image id to build a new cache.

* Spack concretizer unify.

* Rename GH CI test.

* Create new local-buildcache and push to it.

Steven-Roberts committed:
SUNDIALS updates (#105)

* Improvements for IDA and KINSOL

* Remove N_VGetArrayPointer_Serial calls

* Resolve t_init_ issue

* Add missing N_VDestroy

* Apply pre-commmit fixes

* Fix some adjoint memory leaks

* Remove unused function

* Minor suggestions to SUNDIALS interface (#111)

- Create "symmetric" functions for creating and deleting SUNDIALS objects.
- Modify RL example to test Ida destructor.


---------

Co-authored-by: pelesh <pelesh@users.noreply.github.com>

---------

Co-authored-by: Steven-Roberts <Steven-Roberts@users.noreply.github.com>
Co-authored-by: pelesh <peless@ornl.gov>
Co-authored-by: Nicholson Koukpaizan <koukpaizannk@ornl.gov>
Co-authored-by: pelesh <pelesh@users.noreply.github.com>

pelesh committed:
Refactor examples (#110)

* Separate MatPower from other test utilities.

* Update figure path in power flow example README file.

* Fix warning coming from Variable class.

* Fix warnings in Matpower tests.

* Reduce output from phasor dynamics Example 1

* Minor fixes to DG and microgrid examples.


---------

Co-authored-by: pelesh <pelesh@users.noreply.github.com>

pelesh committed:
Move `PowerSystemsData.hpp` and `CircuitGraph.hpp` to correct subdirectories (#114)

* Mowe PowerSystemsData.hpp to src/Models/PowerFlow.

* Move CircuitGraph.hpp to Model/PowerElectronics

* Move Matpower parser to src/Model/PowerFlow

---------

Co-authored-by: pelesh <pelesh@users.noreply.github.com>

alexander-novo committed:
Only check c/c++ files with clang-format pre-commit (#121)

alexander-novo committed:
Update required sundials version in README (#118)

alexander-novo committed:
Add dev container (#86)

* Add dev container

* Update contributing.md with container info

* Reduce container size a bit

Don't install "recommended" packages through `apt`. Also install `make` since it is recommended by `cmake`.

---------

Co-authored-by: alexander-novo <alexander-novo@users.noreply.github.com>

superwhiskers committed:
fix a memory leak, add asan / ubsan support to the cmake configuration (#117)

* add `.editorconfig`, add undefined behavior and address sanitizer support to the cmake build, and fix a memory leak in `DistributedGeneratorTest`

* swap the heap allocated `DistributedGenerator` for one on the stack

* add warnings by default and fix warnings resulting from our code so that it builds cleanly

pelesh committed:
Create phasor dynamics model data structures (#116)

* Create phasor dynamics bus and branch data structures.

* Add load data structure.

* Use BusData with BusInfinite.

* Renamed PowerSystemData -> PowerFlowData


---------

Co-authored-by: pelesh <pelesh@users.noreply.github.com>

pelesh committed:
Create README file documenting classical generator model (#122)

* Creat README file for classical generator model.

superwhiskers committed:
fix remaining warnings (on-repo) (#125)


---------

Co-authored-by: superwhiskers <superwhiskers@users.noreply.github.com>

nkoukpaizan committed:
Reorganization in preparation for Enzyme automatic differentiation PR (#126)

pelesh committed:
Add classical generator model (#127)

abirchfield reviewed (changes requested):
Documentation needs to be updated to reflect structure in #127 and #128 .

I suggest documenting both piecewise and an initial smoothed approximation for the limits.

abirchfield commented:
Also, name of model should be TGOV1 or Governor_TGOV1. TurbineGov could apply to other future governors and we want each to be unique.

pelesh commented:
> Also, name of model should be TGOV1 or Governor_TGOV1. TurbineGov could apply to other future governors and we want each to be unique.

Per developer guidelines class names should be upper camel case format, something like this: `GovernorTgov1`. Would that be acceptable to domain specialists?

We could also use just `Tgov1` and put it in a `Governor` namespace. In that case one would reference the class as `Governor::Tgov1`.

CC @superwhiskers @nkoukpaizan @alexander-novo

abirchfield commented:
GovernorTgov1 is ok.

abirchfield committed:
Changes to compile with MSVC (#130)

abirchfield committed:
Update modeling documentation structure (#128)

* Initial update to GENROU docs

* Initial update to GENSAL docs

* Initial update to load docs

* Initial BusFault docs

* Initial branch docs update

* Correct load math markers

---------

Co-authored-by: abirchfield <abirchfield@users.noreply.github.com>
Co-authored-by: pelesh <peless@ornl.gov>

pelesh committed:
Add component data structures (#133)

superwhiskers committed:
some initial tweaks of the input format (#132)

* update src/Model/PhasorDynamics/INPUT_FORMAT.md

Co-authored-by: pelesh <peless@ornl.gov>

* Aligning with new field labeling paradigm and bus vs node notation

* explicitly specify a date format for consistency, minor adjustments to wording

* indicate default values for bus classes, allow datetimes in the `case_datetime` field

* refine language regarding the datetime field

---------

Co-authored-by: pelesh <peless@ornl.gov>
Co-authored-by: Adam Birchfield <adam.b.birchfield@gmail.com>

superwhiskers committed:
fix several memory leaks (#134)

---------

Co-authored-by: superwhiskers <superwhiskers@users.noreply.github.com>
Co-authored-by: pelesh <peless@ornl.gov>

nkoukpaizan committed:
Enzyme Jacobian for PhasorDynamics::Load within GridKit (#131)

* Separated out Enzyme macros from FindEnzyme and added a EnzymeAddLibrary macro.

* Enzyme wrapper within GridKit.

* Fixes to rebase.

* More interesting vector residual for Enzyme examples.

* Fix some warnings.

* Address one more warning.

* Apply pre-commmit fixes

* Empty commit to trigger CI after pre-commit fixes.

* Successful use of target_compiile_options to pass enzyme-auto-sparsity flags in examples.

* More flags needed to get the correct answer to machine precision in EnzymePowerElectronicsCheck.

* EnzymeAddLibrary not currently needed.

* Apply pre-commmit fixes

* Empty commit to trigger CI after pre-commit fixes.

* Updated documentation in EnzymeAddLibrary.

* Fix warnings previously hidden by manual builds through enzyme_add_executable CMake macro.

* Use GridKit::Testing::isEqual in Enzyme examples.

* Fixes to rebase.

* Apply pre-commmit fixes

* Empty commit to trigger CI after pre-commit fixes.

* residual_wrapper can now be placed within GridKit::Enzyme namespace.

---------

Co-authored-by: nkoukpaizan <nkoukpaizan@users.noreply.github.com>

pelesh committed:
Implement system model data structure and system composer (#138)

* Add bus factory to phasor dynamics family.

* Add system model data for phasor dynamics family.

* Use SystemModelData to instantiate system in phasor dynamic Example 1.

* Add unit test for system constructor.

---------

Co-authored-by: pelesh <pelesh@users.noreply.github.com>

pelesh committed:
Add 10 generator system to phasor dynamics examples (#135)


---------

Co-authored-by: abirchfield <adam.b.birchfield@gmail.com>

lukelowry committed:
chore: environment test gov implementation [skip ci]

lukelowry committed:
chore: environment test gov implementation [skip ci]

lukelowry committed:
edit: component inheritance [skip ci]

lukelowry committed:
edit: component inheritance [skip ci]

lukelowry committed:
edit: turbine compilation issue [skip ci]

lukelowry committed:
edit: turbine compilation issue [skip ci]

lukelowry committed:
init: example 3 turbine include

lukelowry committed:
init: example 3 turbine include

lukelowry committed:
edit: joint genrou and ieeet1 attempt [skip ci]

lukelowry committed:
edit: joint genrou and ieeet1 attempt [skip ci]

lukelowry committed:
fix: compilation error [skip ci]

lukelowry committed:
fix: compilation error [skip ci]

lukelowry committed:
fix: compile issues [skip ci]

lukelowry committed:
fix: compile issues [skip ci]

lukelowry committed:
edit: cmake issue gov [skip ci]

lukelowry committed:
edit: cmake issue gov [skip ci]

lukelowry committed:
edit: type alignments

lukelowry committed:
edit: type alignments

lukelowry committed:
edit: semicolon [skip ci]

lukelowry committed:
fix: cmake correction [skip ci]

lukelowry committed:
fix: cmake correction [skip ci]

lukelowry committed:
edit: method signature [skip ci]

lukelowry committed:
edit: method signature [skip ci]

lukelowry committed:
edit: constructor overloading

lukelowry committed:
edit: constructor overloading

lukelowry committed:
edit: compile error [skip ci]

lukelowry committed:
edit: compile error [skip ci]

lukelowry committed:
fix: example runtime [skip ci]

lukelowry committed:
fix: example runtime [skip ci]

pelesh commented:
Please make sure you rename the example. We already have `Example3`, it's the ten-gen model.

lukelowry committed:
chore: new doc format and misc improvments [skip ci]

lukelowry committed:
chore: new doc format and misc improvments [skip ci]

lukelowry committed:
Apply pre-commmit fixes

lukelowry committed:
Merge branch 'lukel/phasor_tgov1_dev' of https://github.com/ORNL/GridKit into lukel/phasor_tgov1_dev

lukelowry committed:
edit: rebase correction [skip ci]

lukelowry committed:
Merge branch 'lukel/phasor_tgov1_dev' of https://github.com/ORNL/GridKit into lukel/phasor_tgov1_dev

lukelowry committed:
edit: misc updates from rebase [skip ci]

lukelowry committed:
edit: SystemModel capitalization issue [skip ci]

lukelowry committed:
edit: example 3 rebase fix [skip ci]

lukelowry committed:
edit: example 3 rebase fix [skip ci]

lukelowry committed:
add: TurbineGovData [skip ci]

lukelowry committed:
Apply pre-commmit fixes

lukelowry committed:
add: Example 4 Init

lukelowry committed:
edit: Gov data format error [skip ci]

lukelowry committed:
edit: struct definition correction [skip ci]

lukelowry committed:
Apply pre-commmit fixes

lukelowry committed:
edit: example 4 model init [skip ci]

lukelowry committed:
edit: fix type parameter

lukelowry committed:
Apply pre-commmit fixes

lukelowry committed:
edit: allocation example 4 [skip ci]

lukelowry committed:
Merge branch 'lukel/phasor_tgov1_dev' of https://github.com/ORNL/GridKit into lukel/phasor_tgov1_dev

lukelowry committed:
Apply pre-commmit fixes

lukelowry committed:
add: gov setter for gen

lukelowry committed:
Merge branch 'lukel/phasor_tgov1_dev' of https://github.com/ORNL/GridKit into lukel/phasor_tgov1_dev

lukelowry committed:
fix: Gov data struct defined incorrectly

lukelowry committed:
Merge branch 'lukel/phasor_tgov1_dev' of https://github.com/ORNL/GridKit into lukel/phasor_tgov1_dev

lukelowry committed:
edit: cmake example 4 link turbine gov [skip ci]

lukelowry committed:
edit: typical gov params (docs and example) [skip ci]

lukelowry committed:
edit: initialization correction w no limit function [skip ci]

lukelowry committed:
add: smooth indicator implementation [skip ci]

lukelowry committed:
edit: valve sign deriv [skip ci]

lukelowry committed:
edit: parameter tuning [skip ci]

lukelowry committed:
fix: IVP correction [skip ci]

lukelowry committed:
fix: incorrect override keyword [skip ci]

lukelowry committed:
add: MachineBase cirtual torque access [skip ci]

lukelowry committed:
edit: fault time to assess IVP stabilize [skip ci]

lukelowry committed:
edit: modify torque access [skip ci]

lukelowry committed:
edit: time of fault change caused segmentation fault [skip ci]

lukelowry committed:
edit: increase valve upper bound example 4 [skip ci]

lukelowry committed:
edit: sign of activation valve limits [skip ci]

lukelowry committed:
edit: pmech sign error [skip ci]

lukelowry committed:
edit: pass pmech by reference [skip ci]

lukelowry committed:
edit: state order [skip ci]

lukelowry committed:
checkpoint: valid simulation minimal error [skip ci]

lukelowry committed:
add: example 4 PW reference [skip ci]

lukelowry committed:
edit: camel case governor [skip ci]

lukelowry committed:
edit: file path correction [skip ci]

lukelowry committed:
edit: more file path correction [skip ci]

lukelowry committed:
edit: misc documentation formatting [skip ci]

lukelowry committed:
add: unit test environment [skip ci]

lukelowry committed:
edit: polymorph correction [skip ci]

lukelowry committed:
edit: descructor of Machine [skip ci]

lukelowry committed:
edit: unit test cmake error

lukelowry committed:
edit: unit test cmake
